File: bot-api/dotnet/Robocode.TankRoyale.BotApi/src/events/HitByBulletEvent.cs
using Newtonsoft.Json;

namespace Robocode.TankRoyale.BotApi.Events;

/// <summary>
/// Event occurring when a bullet has hit your bot.
/// </summary>
public sealed class HitByBulletEvent : BotEvent
{
    /// <summary>Bullet that hit the bot.</summary>
    public BulletState Bullet { get; }

    /// <summary>Damage inflicted by the bullet.</summary>
    public double Damage { get; }

    /// <summary>Remaining energy level of the bot that got hit.</summary>
    public double Energy { get; }

    /// <summary>
    /// Initializes a new instance of the BulletHitBotEvent class.
    /// </summary>
    /// <param name="turnNumber">Turn number.</param>
    /// <param name="bullet">Bullet that hit the bot.</param>
    /// <param name="damage">Damage inflicted by the bullet.</param>
    /// <param name="energy">Remaining energy level of the bot that got hit.</param>
    [JsonConstructor]
    public HitByBulletEvent(int turnNumber, BulletState bullet, double damage, double energy) : base(turnNumber) =>
        (Bullet, Damage, Energy) = (bullet, damage, energy);
}

File: bot-api/typescript/src/events/HitByBulletEvent.ts
import { BotEvent } from "./BotEvent.js";
import { BulletState } from "../BulletState.js";

/** Event occurring when a bot has been hit by a bullet. */
export class HitByBulletEvent extends BotEvent {
  readonly bullet: BulletState;
  readonly damage: number;
  readonly energy: number;

  constructor(turnNumber: number, bullet: BulletState, damage: number, energy: number) {
    super(turnNumber);
    this.bullet = bullet;
    this.damage = damage;
    this.energy = energy;
  }
}
